### PR TITLE
Add Support for Vault Backup & Restore

### PR DESCRIPTION
**Pull request by sakiib**
Signed-off-by: sakibalamin <sakibalamin@appscode.com>

### CODE REVIEW

File: pkg/backup.go [hunk 1]
```diff
@@ -192,41 +203,148 @@ func (opt *vaultOptions) backupVault(targetRef api_v1beta1.TargetRef) (*restic.B
...
+	err = opt.saveVaultSnapshot(session)
+	if err != nil {
+		return nil, err
+	}
+
+	// save the unseal keys & root token too, for complete backup
+	// write these in the interim directory along with the snapshot file
+	if err := opt.writeVaultTokenKeys(vs); err != nil {
+		return nil, err
+	}
```
Comment (hmsayem) [inline pkg/backup.go:273]: can we use the following format for returing from these methods?

```go
if err := anyfunc(); err != nil {
    return nil, err
}
```

File: pkg/backup.go [hunk 2]
```diff
@@ -192,41 +203,148 @@ func (opt *vaultOptions) backupVault(targetRef api_v1beta1.TargetRef) (*restic.B
...
+	for _, key := range keys {
+		value, err := st.Get(key)
+		if err != nil {
+			klog.Errorf("failed to get key %s with %s\n", key, err.Error())
+			return err
+		}
+
+		if err := opt.write(key, value); err != nil {
+			klog.Errorf("failed to write key %s with %s\n", key, err.Error())
+		}
```
Comment (hmsayem) [inline pkg/backup.go:333]: will it affect the restore process if a key is not written? If yes. Should we return the error from here?

File: pkg/backup.go [hunk 3]
```diff
@@ -192,41 +203,148 @@ func (opt *vaultOptions) backupVault(targetRef api_v1beta1.TargetRef) (*restic.B
...
+	return nil
+}
+
+func (opt *VaultOptions) write(key, value string) error {
+	byteStreams, err := json.Marshal(value)
+	if err != nil {
+		return err
+	}
+
+	if err = os.WriteFile(filepath.Join(opt.interimDataDir, key), byteStreams, 0o644); err != nil {
```
Comment (hmsayem) [inline pkg/backup.go:346]: can we directly return function?
```go
return os.WriteFile(filepath.Join(opt.interimDataDir, key), byteStreams, 0o644)
```

File: pkg/restore.go [hunk 4]
```diff
@@ -156,40 +174,167 @@ func (opt *vaultOptions) restoreVault(targetRef api_v1beta1.TargetRef) (*restic.
...
-	// append the restore command to the pipeline
-	opt.dumpOptions.StdoutPipeCommands = append(opt.dumpOptions.StdoutPipeCommands, *session.cmd)
-	resticWrapper, err := restic.NewResticWrapperFromShell(opt.setupOptions, session.sh)
+	// set the vault connection parameters, essentially the vault leader node address
+	err = session.setVaultConnectionParameters(vaultClient, appBinding)
 	if err != nil {
 		return nil, err
 	}
-	return resticWrapper.Dump(opt.dumpOptions, targetRef)
+
```
Comment (hmsayem) [inline pkg/restore.go:231]: can we use the following format for returing from these methods?

```go
if err := anyfunc(); err != nil {
    return nil, err
}
```

File: pkg/restore.go [hunk 5]
```diff
@@ -156,40 +174,167 @@ func (opt *vaultOptions) restoreVault(targetRef api_v1beta1.TargetRef) (*restic.
...
+	}
+
+	restoreOutput, err := resticWrapper.RunRestore(opt.restoreOptions, targetRef)
+	if err != nil {
+		return nil, err
+	}
+
+	// restore the vault snapshot from the interim directory running command like: vault operator raft snapshot restore backup.snap
+	// for different vault server using -force flag like: vault operator raft snapshot restore -force backup.snap
+	err = opt.restoreVaultSnapshot(session)
```
Comment (hmsayem) [inline pkg/restore.go:248]: same as previous

File: pkg/restore.go [hunk 6]
```diff
@@ -156,40 +174,167 @@ func (opt *vaultOptions) restoreVault(targetRef api_v1beta1.TargetRef) (*restic.
...
+
+	for _, key := range keys {
+		value, err := opt.read(key)
+		if err != nil {
+			klog.Errorf("failed to read key %s with %s\n", key, err.Error())
+			return err
+		}
+
+		if err := st.Set(key, value); err != nil {
+			klog.Errorf("failed to set key %s with %s\n", key, err.Error())
```
Comment (hmsayem) [inline pkg/restore.go:321]: will it affect the backup process if a key is not set?

**Review by hmsayem — CHANGES_REQUESTED** [review]

File: pkg/restore.go (hunk 6 above)
Comment (sakiib) [reply pkg/restore.go:321]: yes, these are some valid points, let me update the PR!

File: pkg/utils.go [hunk 7]
```diff
@@ -81,37 +99,33 @@ func (opt *vaultOptions) newSessionWrapper(cmd string) *sessionWrapper {
...
+	tokenSecret, err := kubeClient.CoreV1().Secrets(appBinding.Namespace).Get(context.TODO(), vs.BackupSecretName(), metav1.GetOptions{})
 	if err != nil {
 		return err
 	}
 
-	err = appBinding.TransformSecret(kubeClient, appBindingSecret.Data)
+	err = appBinding.TransformSecret(kubeClient, tokenSecret.Data)
 	if err != nil {
 		return err
 	}
```
Comment (hmsayem) [inline pkg/utils.go:111]:
```go
if err = appBinding.TransformSecret(kubeClient, tokenSecret.Data);  err != nil {
    return err
}
```

File: pkg/backup.go [hunk 8]
```diff
@@ -147,11 +150,13 @@ func NewCmdBackup() *cobra.Command {
...
 
 	cmd.Flags().StringVar(&opt.outputDir, "output-dir", opt.outputDir, "Directory where output.json file will be written (keep empty if you don't need to write output in file)")
+	cmd.Flags().StringVar(&opt.interimDataDir, "interim-data-dir", opt.interimDataDir, "Directory where the targeted data will be stored temporarily before uploading to the backend")
+	cmd.Flags().StringVar(&opt.keyPrefix, "key-prefix", opt.keyPrefix, "prefix that will be append to root-token & unseal-keys")
 
 	return cmd
 }
 
-func (opt *vaultOptions) backupVault(targetRef api_v1beta1.TargetRef) (*restic.BackupOutput, error) {
+func (opt *VaultOptions) backupVault(targetRef api_v1beta1.TargetRef) (*restic.BackupOutput, error) {
```
Comment (hossainemruz) [inline pkg/backup.go:159]: Make `VaultOptions` unexported.

File: pkg/backup.go [hunk 9]
```diff
@@ -192,41 +195,140 @@ func (opt *vaultOptions) backupVault(targetRef api_v1beta1.TargetRef) (*restic.B
...
 	if err != nil {
 		return nil, err
 	}
 
-	session := opt.newSessionWrapper(VaultDumpCMD)
+	parameters := vaultconfig.VaultServerConfiguration{}
+	if appBinding.Spec.Parameters != nil {
+		if err = json.Unmarshal(appBinding.Spec.Parameters.Raw, &parameters); err != nil {
+			klog.Errorf("unable to unmarshal appBinding.Spec.Parameters.Raw. Reason: %v", err)
+		}
```
Comment (hossainemruz) [inline pkg/backup.go:208]: Change to
```
return fmt.Errorf("unable to unmarshal appBinding.Spec.Parameters.Raw: %w", err)
```

File: pkg/backup.go [hunk 10]
```diff
@@ -192,41 +195,140 @@ func (opt *vaultOptions) backupVault(targetRef api_v1beta1.TargetRef) (*restic.B
...
+	}
 
-	err = session.setDatabaseCredentials(opt.kubeClient, appBinding)
-	if err != nil {
+	// update this while adding support for more backend options for backup (consul, s3, etc.)
+	if parameters.Backend != VaultStorageBackendRaft {
+		return nil, errors.New("Backend must be Raft for backup snapshots")
```
Comment (hossainemruz) [inline pkg/backup.go:213]: Use `fmt.Errorf()`.

File: pkg/backup.go [hunk 11]
```diff
@@ -192,41 +195,140 @@ func (opt *vaultOptions) backupVault(targetRef api_v1beta1.TargetRef) (*restic.B
...
+	var keys []string
+	keys = append(keys, opt.tokenName(opt.keyPrefix))
+	for i := 0; i < int(params.Unsealer.SecretShares); i++ {
+		keys = append(keys, opt.unsealKeyName(opt.keyPrefix, i))
+	}
+
+	for _, key := range keys {
+		value, err := st.Get(key)
+		if err != nil {
+			klog.Errorf("failed to get key %s with %s\n", key, err.Error())
```
Comment (hossainemruz) [inline pkg/backup.go:313]: Don't log if you return the error. Just return instead.
```go
return fmt.Errorf("failed to get key %s: %w", key, err)
```

File: pkg/restore.go [hunk 12]
```diff
@@ -156,40 +170,168 @@ func (opt *vaultOptions) restoreVault(targetRef api_v1beta1.TargetRef) (*restic.
...
+	// for different vault server using -force flag like: vault operator raft snapshot restore -force backup.snap
+	if err := opt.restoreVaultSnapshot(session); err != nil {
+		return nil, err
+	}
+
+	// if -force is true, we're potentially dealing with restoring snapshot on a different vault server
+	// we must replace the current vault server's unseal keys & root token with the older ones that we saved during snapshot
+	if opt.force {
+		klog.Infoln("Potentially different VaultServer. Trying to migrate old unseal keys & root token.")
+		if err := opt.setVaultTokenKeys(appBinding, parameters); err != nil {
```
Comment (hossainemruz) [inline pkg/restore.go:249]: You can rename `setVaultTokenKeys()` to `migrateVaultTokenKeys()` `overwriteVaultTokensKeysWithOldTokens()`

File: pkg/restore.go [hunk 13]
```diff
@@ -156,40 +170,168 @@ func (opt *vaultOptions) restoreVault(targetRef api_v1beta1.TargetRef) (*restic.
...
+	newKeys = append(newKeys, opt.tokenName(opt.keyPrefix))
+	for i := 0; i < int(params.Unsealer.SecretShares); i++ {
+		newKeys = append(newKeys, opt.unsealKeyName(opt.keyPrefix, i))
+	}
+
+	for idx, oldKey := range oldKeys {
+		value, err := opt.read(oldKey)
+		if err != nil {
+			klog.Errorf("failed to read key %s with %s\n", oldKey, err.Error())
+			return err
```
Comment (hossainemruz) [inline pkg/restore.go:313]:
```go
return fmt.Errorf("failed to read key %s. Reason: %w", oldKey, err)
```

File: pkg/store/azure/azure.go [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+
+	vaultBaseUrl := store.azureSpec.VaultBaseURL
+	client := azsecrets.NewClient(vaultBaseUrl, store.cred, nil)
+
+	_, err := client.SetSecret(context.TODO(), key, azsecrets.SetSecretParameters{
+		Value:       pointer.StringP(base64.StdEncoding.EncodeToString([]byte(value))),
+		ContentType: pointer.StringP("password"),
+	}, nil)
+	if err != nil {
+		return errors.Wrap(err, "unable to set secrets in key vault")
```
Comment (hossainemruz) [inline pkg/store/azure/azure.go:123]: Use `fmt.Errorf("unable to set secrets in key vault: %w")`

**Review by hossainemruz — CHANGES_REQUESTED** [review]

**Review by hossainemruz — APPROVED** [review]

**Pull request merged**